src/annotator/highlighter.js

```diff
@@ -161,17 +161,37 @@ function wholeTextNodesInRange(range) {
     return [];
   }
 
-  const textNodes = [];
+  /** @type {Node|null} */
+  let root = range.commonAncestorContainer;
+  if (root.nodeType !== Node.ELEMENT_NODE) {
+    // If the common ancestor is not an element, set it to the parent element to
+    // ensure that the loop below visits any text nodes generated by splitting
+    // the common ancestor.
+    root = root.parentElement;
+  }
+  if (!root) {
+    // If there is no root element then we won't be able to insert highlights,
+    // so exit here.
+    return [];
+  }
 
-  forEachNodeInRange(range, node => {
-    if (node.nodeType !== Node.TEXT_NODE) {
-      return;
+  const textNodes = [];
+  const nodeIter = /** @type {Document} */ (root.ownerDocument).createNodeIterator(
+    root,
+    NodeFilter.SHOW_TEXT
+  );
+  let node;
+  while ((node = nodeIter.nextNode())) {
+    if (!isNodeInRange(range, node)) {
+      continue;
     }
     let text = /** @type {Text} */ (node);
 
     if (text === range.startContainer && range.startOffset > 0) {
-      // Split `text` where the range starts, setting it to the part in the range.
-      text = text.splitText(range.startOffset);
+      // Split `text` where the range starts. The part that is in the range will
+      // be visited in the next loop iteration.
+      text.splitText(range.startOffset);
```

Comment:
There's some magic in `Text.createText` that I think should be commented about here. From MDN, 

> After the split, the current node contains all the content up to the specified offset point, and a newly created node of the same type contains the remaining text. The newly created node is returned to the caller. If the original node had a parent, the **new node is inserted as the next sibling of the original node**.

(emphasis mine)

The magic being that the newly-created `Text` node (containing the text that _is_ in the `Range`, IIUC) will be inserted, and _that_ (I think?) is how it's caught in further iterations?

Reply:
Yes, exactly. I've added a comment to clarify.